HP-380 fix vulnerabilities

NikoHelle opened the pull request:
Fixes vulnerabilities found by Snyk: https://app.snyk.io/org/city-of-helsinki/project/166fa581-33c0-4bfc-90ba-513e2bf8bf94/

This PR updates vulnerable packages which have an updated version. Some updates were major version changes, so quite many changes had to be made after the update. But mostly fixes for linting errors raised by react-scripts update. Two test files needed actual code changes to fix tests.

Sadly this PR wont fix all vulnerabilities because there are no updates to address new vulnerabilities in updated packages.

NikoHelle committed:
Updated required packages and their dependencies

Changed packages:

- yup had severe vulnerabilites

- react-scripts had most vulnerabilies and required major version update

- i18n needed update too and updated also other x-i18n packages for compatibility

- typescript was also required to update because of i18n and new "import type {...}

- lodash had a severe vulnerability

- added also eslint to devDependencies because multiple packages require it

- had to add eslint-config-react-app to dependencies because of yarn bug which installs it with react-scripts under react-script -folder where it is not found.

Other changes:

Removed eslintConfig from package.json because it is in .eslintrc

Changes in tsconfig.json are made by react-scripts

Fixed all error-level parsing errors in JS/TS files by adding rule overrides.

NikoHelle committed:
Fix lint errors introduced in react-script & typescript update

Global eslint changes in .eslintrc:

- no-use-before-define was added / fixed as stated in:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-use-before-define.md#how-to-use

- "no-undef": "error" was added to prevent compile errors.

Other changes in tsx/ts -files fix mostly linting errors when function return type is not defined or module exports something by default which is not named first.

Obsolete jest-localstorage-mock was removed because new version of Jest includes same features and they do not work together.

NikoHelle committed:
Fix broken tests

New localstorage mock from Jest caused spy-functions to break

Toast.tsx broke because it really didnt work before and updated Jest caused an error when trying to remove element that is not found.

NikoHelle committed:
Remove disabled linting and fix raised errors

Some linting errors where disabled with eslint-disable-line. Removed those and fixed errors caused by them.

Removed also one useless line that was commented out

NikoHelle commented on .gitignore:
```diff
@@ -21,3 +21,4 @@
 npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
+.eslintcache
```
.eslintcache is added by react-scripts v4

NikoHelle commented on package.json:
```diff
@@ -81,11 +78,13 @@
   },
   "devDependencies": {
     "apollo": "^2.21.1",
+    "eslint": "^7.15.0",
```
Added because so many dependencies require it on their own.

NikoHelle commented on src/common/header/userDropdown/UserDropdown.tsx:
```diff
@@ -12,6 +12,7 @@ import useToast from '../../../toast/useToast';
 const NAME_QUERY = loader('../../../profile/graphql/NameQuery.graphql');
 
+// eslint-disable-next-line @typescript-eslint/ban-types
 type Props = {};
```
These ban-types are removed in another commit.

NikoHelle commented on src/auth/__tests__/authService.test.js:
```diff
@@ -25,9 +24,10 @@ describe('authService', () => {
       authService.getToken();
 
-      expect(localStorage.getItem).toHaveBeenNthCalledWith(3, API_TOKEN);
+      expect(getSpy).toHaveBeenNthCalledWith(1, API_TOKEN);
```
It does not makes sense why getItem would be called 3 times in this test. Maybe previous spy did not reset properly between tests.

NikoHelle commented on src/toast/__tests__/Toast.test.jsx:
```diff
@@ -44,7 +44,7 @@ describe('Toast', () => {
   afterAll(() => {
-    global.document.getElementById('toast').remove();
+    global.document.getElementById('toast-root').remove();
   });
```
This test did not make sense either. There is no element with id 'toast'. Previous version of JSDOM may have worked more like jQuery and returned callable functions when query does not return anything. But now calling remove() threw and error.

Also there is only one test, so unsuccessful removal in afterAll() have no effect to following tests.

akikoskinen commented on src/toast/Toast.tsx:
```diff
@@ -6,13 +6,20 @@ import { ToastTypes } from './types';
   title?: string;
   description?: string;
+  // eslint-disable-next-line react/no-unused-prop-types
   id: string;
```
Is `id` needed at all here, if it's not used?

NikoHelle replied:
It was needed due the way Toast.tsx is used: by passing Toast object which has the id. This change is now outdated, because I changed the component to use Toast interface.

akikoskinen reviewed (approved):
Looks good. One small question.

NikoHelle committed:
Fix typescript error while running docker-compose up. @typescript-eslint/ban-ts-comment not recognised

NikoHelle committed:
Fix ts compile error with useProfile.ts and missing prop error caused by outdated @types

NikoHelle committed:
Extend Toast interface in Toast React component

Component props almost matched Toast interface so extending the interface is better than defining same props again

NikoHelle committed:
Move eslint*-* to dependencies to avoid docker build error

NikoHelle committed:
Removed useless eslint-config-react-app

It was added in one commit to prevent lint error that is no longer happening

akikoskinen commented:
Still looking good.
